actionpack/lib/action_dispatch/http/content_security_policy.rb

```diff
@@ -361,5 +420,26 @@ def resolve_source(source, context)
       def nonce_directive?(directive, nonce_directives)
         nonce_directives.include?(directive)
       end
+
+      def build_report_directives(proc, report_to_endpoints, reporting_endpoints)
+        csp_report_endpoints = proc.call
+        unless csp_report_endpoints.is_a?(Hash)
+          raise ArgumentError, "Invalid CSP reporting endpoints from Proc: #{csp_report_endpoints.inspect}. Must return a Hash."
+        end
+        csp_report_endpoints.each do |key, endpoint|
+          case endpoint
+          when String
+            report_uri(endpoint)
+            reporting_endpoints << "#{key}=\"#{endpoint}\""
+          when Hash
+            urls = endpoint.delete(:urls) || []
+            endpoint[:group] = key
+            endpoint[:endpoints] = urls.filter_map { |url| { url: url } unless url.nil? }
+            report_to_endpoints << endpoint.to_json
+          else
+            raise ArgumentError, "Invalid CSP reporting endpoint type: #{endpoint.class}. Must be String or Hash."
```

Comment:
Just a plain argument error with the type is not very useful, how about including the provided data as well?